File: projectpilot/src/BussinessLogic/Models/Generics/PagedResults.ts
export interface PagedResults<T> {
    data: T[];
    currentPage: number;
    totalRecords: number;
    totalPages: number;
    pageSize: number;
    nextPage?: number;
    prevPage?: number;
}


export function createPagedResults<T>(raw: any, modelFactory?: (raw: any) => T) {

    return {
        currentPage: raw.currentPage ?? 1,
        totalRecords: raw.totalRecords ?? 1,
        totalPages: raw.totalPages ?? 1,
        pageSize: raw.pageSize ?? 5,
        nextPage: raw.nextPage,
        prevPage: raw.prevPage,
        data: mapData<T>(raw.data, modelFactory)

    };
}

export function createDefaultPagedResults() {

    return {
        currentPage: 1,
        totalRecords: 1,
        totalPages: 1,
        pageSize: 5,
        nextPage: undefined,
        prevPage: undefined,
        data: []

    };
}

function mapData<T>(raw: any, modelFactory?: (raw: any) => T) {
    return modelFactory && Array.isArray(raw) ?
        raw.map((item: any) => modelFactory(item)) : [];
}

File: FmdcOperations/Domain/Helpers/PagedResults.cs
namespace Domain.Helpers
{
    public class PagedResults<T> where T : class
    {
        public IEnumerable<T> Data { get; set; } = [];
        public int CurrentPage { get; set; } = 1;
        public int TotalRecords { get; set; }
        public int TotalPages { get; set; }
        public int PageSize { get; set; }
        public int? NextPage { get; set; }
        public int? PrevPage { get; set; }

        public PagedResults(IEnumerable<T> data, int currentPage, int totalRecords, int pageSize)
        {
            Data = data;
            CurrentPage = currentPage;
            TotalRecords = totalRecords;
            PageSize = pageSize;
            TotalPages = (int)Math.Ceiling((double)totalRecords / pageSize);
            NextPage = CurrentPage < TotalPages ? CurrentPage + 1 : (int?)null;
            PrevPage = CurrentPage > 1 ? CurrentPage - 1 : (int?)null;
        }
    }

}
